sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,247 @@
+		this.#cache.delete(key);
+	}
+
+	async clear(): Promise<void> {
+		this.#cache.clear();
+	}
+
+	static createCache<T>(): AsyncCache<T> {
+		return new InMemoryCache<T>();
+	}
+}
+
+interface ObjectCacheOptions {
+	createCache?: <T>(name: string) => AsyncCache<T>;
+}
+
+export class ObjectCache implements TransactionBlockDataResolverPlugin {
+	#createCache: <T>(name: string) => AsyncCache<T>;
+	#objects: AsyncCache<ObjectCacheEntry>;
+	#functions: AsyncCache<MoveFunctionEntry>;
+
+	constructor({ createCache = InMemoryCache.createCache }: ObjectCacheOptions = {}) {
+		this.#createCache = createCache;
+		this.#objects = this.#createCache('objects');
+		this.#functions = this.#createCache('functions');
+	}
+
+	getObjects: NonNullable<TransactionBlockDataResolverPlugin['getObjects']> = async (ids, next) => {
+		const results = new Map<string, ObjectCacheEntry>();
+
+		await Promise.all(
+			ids.map(async (id) => {
+				const cachedObject = await this.#objects.get(id);
+
+				if (cachedObject) {
+					results.set(id, cachedObject);
+				}
+			}),
+		);
+		const missingIds = ids.filter((id) => !results.has(id));
+
+		if (missingIds.length > 0) {
+			const newObjects = await next(missingIds);
+
+			await Promise.all(
+				newObjects.map(async (newObject) => {
+					await this.addObject(newObject);
+					results.set(newObject.objectId, newObject);
+				}),
+			);
+		}
+
+		return ids.map((id) => results.get(id)!);
+	};
+
+	getMoveFunctionDefinition: NonNullable<
+		TransactionBlockDataResolverPlugin['getMoveFunctionDefinition']
+	> = async (ref, next) => {
+		const functionName = `${normalizeSuiAddress(ref.package)}::${ref.module}::${ref.function}`;
+		const cached = await this.#functions.get(functionName);
+		if (cached) {
+			return cached;
+		}
+
+		const functionDefinition = await next(ref);
+
+		return await this.addFunction(functionDefinition);
+	};
+
+	async clearCache() {
+		await Promise.all([this.#objects.clear(), this.#functions.clear()]);
+	}
+
+	async addObject(object: ObjectCacheEntry) {
+		await this.#objects.set(object.objectId, object);
+		return object;
+	}
+
+	async addFunction(functionEntry: MoveFunctionEntry) {
+		const pkg = normalizeSuiAddress(functionEntry.package);
+		const functionName = `${pkg}::${functionEntry.module}::${functionEntry.function}`;
+		const entry = {
+			...functionEntry,
+			package: pkg,
+		};
+
+		await this.#functions.set(functionName, entry);
+
+		return entry;
+	}
+
+	async invalidateObject(id: string) {
+		await this.#objects.delete(id);
+	}
+
+	async applyEffects(effects: typeof bcs.TransactionEffects.$inferType) {
+		if (!effects.V2) {
+			throw new Error(`Unsupported transaction effects version ${effects.$kind}`);
+		}
+
+		const { lamportVersion, changedObjects } = effects.V2;
+
+		await Promise.all(
+			changedObjects.map(async ([id, change]) => {
+				if (change.outputState.NotExist) {
+					await this.invalidateObject(id);
+				} else if (change.outputState.ObjectWrite) {
```

Comment:
Should we be more careful about object owners here?  This just adds all objects to the cache, do we want to filter out objects that end up owned by other addresses?  If an object ends up owned by another address, should we delete it from the cache, or just ignore it?

Reply:
The only owned objects in the transaction should be owned by us, right?

Reply:
yes, thats why I was thinking this is fine, effects will include objects transferred to other addresses, but we should never end up reading those objects because that would imply you are constructing a transaction with objects you don't own, so we could just not store them.  The only edge case is if you somehow send objects somewhere else, then it gets sent back by another transaction, without you clearing the cache, then you try to construct a transaction again with the same object.  I think it's probably more correct to just remove anything from the cache that you don't own after the transaction

Reply:
Yeah I agree, that makes sense to me